lib/Serialization/Deserialization.cpp

```diff
@@ -4060,9 +4061,10 @@ llvm::Error DeclDeserializer::deserializeDeclAttributes() {
         ArrayRef<uint64_t> parameters;
 
         serialization::decls_block::DifferentiableDeclAttrLayout::readRecord(
-            scratch, isImplicit, linear, jvpNameId, jvpDeclId, vjpNameId,
-            vjpDeclId, derivativeGenSigId, parameters);
+            scratch, isImplicit, originalDeclId, linear, jvpNameId, jvpDeclId,
+            vjpNameId, vjpDeclId, derivativeGenSigId, parameters);
 
+        FuncDecl *originalDecl = cast<FuncDecl>(MF.getDecl(originalDeclId));
```

Comment:
The issue is this line causes an infinite loop: we cannot deserialize the original declaration here because `@differentiable` attributes are deserialized as part of deserializing the original declaration.

Reply:
Fixed in https://github.com/apple/swift/pull/27985/commits/4df03077b64db926a03d4551f4ce3599ca4bb54b via ad-hoc deserialization logic. I couldn't think of a better solution.